pkg/volume/util/atomic_writer.go

```diff
@@ -0,0 +1,391 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// AtomicWriter handles atomically projecting content for a set of files into
+// a target directory.  AtomicWriter maintains a sentinel file named
+// "..sentinel" in the target directory which is updated after new data is
+// projected into the target directory, allowing consumers of the data to
+// listen for updates by monitoring the sentinel file with inotify or
+// fanotify.
+//
+// The visible files in this volume are symlinks to files in the writer's data
+// directory.  Actual files are stored in a hidden timestamped directory which
+// is symlinked to by the data directory. The timestamped directory and
+// data directory symlink are created in the writer's target dir.  This scheme
+// allows the files to be atomically updated by changing the target of the
+// data directory symlink.
+type AtomicWriter struct {
+	TargetDir  string
+	LogContext string
+}
+
+func NewAtomicWriter(targetDir, podNamespace, podName, volumeName string) (*AtomicWriter, error) {
+	_, err := os.Stat(targetDir)
+	if os.IsNotExist(err) {
+		return nil, err
+	}
+
+	context := fmt.Sprintf("%v/%v volume: %v", podNamespace, podName, volumeName)
```

Comment:
nothing else in this lib know about pods or volumes - why not just let the user pass in context.